libr/anal/esil_cfg.c

```diff
@@ -0,0 +1,687 @@
+static RAnalEsilCFG *esil_cfg_gen (RAnalEsilCFG *cfg, RAnal *anal, RIDStorage *atoms, RContRBTree *blocks, RStack *stack, ut64 off, char *expr) {
+	// consider expr as RStrBuf, so that we can sanitze broken esil
+	// (ex: "b,a,+=,$z,zf,:=,7,$c,cf,:=,zf,?{,1,b,+=,cf,?{,3,a,-=" =>
+	// 	"b,a,+=,$z,zf,:=,7,$c,cf,:=,zf,?{,1,b,+=,cf,?{,3,a,-=,},}")
+
+	// allocate some stuff
+	char *_expr = strdup(expr);
+	if (!_expr) {
+		return cfg;	//NULL?
+	}
+	RAnalEsilBB *end_bb = R_NEW0(RAnalEsilBB);
+	if (!end_bb) {
+		free(_expr);
+		return cfg;
+	}
+	RGraphNode *start, *end = r_graph_add_node(cfg->g, end_bb);
+	if (!end) {
+		free(end_bb);
+		free(_expr);
+		return cfg;
+	}
+	end->free = _free_bb_cb;
+
+	esil_expr_atomize(atoms, _expr);
+
+	// previous expression's post-dominator is the current expression starting point
+	//
+	// MUST NOT use cfg->start as starting point of subgraph,
+	// since it marks the start of the whole graph
+	//
+	// cpu-instruction starts at this node
+	//
+	// without information about the outside cfg, we CANNOT merge cpu-instructions
+	
+	RAnalEsilBB *bb = (RAnalEsilBB *)cfg->end->data;
+
+	end_bb->expr = bb->expr;
+	// FIXME: use end_bb here
+	bb->expr = NULL;
+	bb->first.off = bb->last.off = off;
+	bb->first.idx = bb->last.idx = 0;
+	start = cfg->end;
+
+	EsilCfgGen gen = {anal->esil, {stack}, blocks, cfg, start, atoms, off};
+	cfg->end = end;
+	// create an edge from cur to end?
+	// Well yes, but no. Would be great to do this,
+	// but rgraph api is slow af on node delition. Be careful instead
+	
+	// We created a new graph node above, which is going to be the post-dominator
+	// of the subgraph, that we are going to add to the existing graph.
+	// The post-dominator of the previous added subgraph is the starting node here.
+	// We add this to the block-tree
+	r_rbtree_cont_insert(blocks, gen.cur, _graphnode_esilbb_insert_cmp, NULL);
+
+	// end of the initial setup, next generate blocks and insert them in the tree
+
+	// round 0 adds a subgraph from if, else and fi
+	r_id_storage_foreach(atoms, _round_0_cb, &gen);
+	// make cfg->end effective post-dominator
+	r_graph_add_edge(cfg->g, gen.cur, cfg->end);
+	{
+		// stack unwinding
+		EsilCfgScopeCookie *cookie;
+		while ((cookie = r_stack_pop(stack))) {
+			r_graph_add_edge(cfg->g, 
+				cookie->is_else ? cookie->if_block : cookie->else_block, cfg->end);
+			free(cookie);
+		}
+	}
+
+	// next do round 1: split blocks from GOTOs and BREAKs
+	r_id_storage_foreach(atoms, _round_1_cb, &gen);
+
+	// next do dfs:
+	//  - remove each node from blocks-tree, that can be reached by a dfs path
+	//  - when removing a node from block-tree, synthesize node->bb->expr with RStrBuf
+	{
+		// dfs walk removes used atoms
+		RGraphVisitor vi = {_round_2_cb, NULL, NULL, NULL, NULL, &gen};
+		r_graph_dfs_node(cfg->g, start, &vi);
+	}
+	// this loop removes unused atoms
+	do {} while (blocks->root && r_rbtree_cont_delete(blocks, NULL, _graphnode_delete_always_0_cmp, &gen));
```

Comment:
missing space before (, and i would prefer to split this into multiple lines for readability